### ISSUE TITLE
Vulkan: DirectionalLight3D shadow rendering can miss long objects, causing shadows to disappear in close splits

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot#60391]
### Godot version

4.0.alpha (1d2177938)

### System information

Fedora 34, GeForce GTX 1080 (NVIDIA 510.60.02)

### Issue description

DirectionalLight3D shadow rendering can miss long objects, causing shadows to disappear in close splits

This occurs regardless of the object's rotation relative to the DirectionalLight3D shadow.

https://user-images.githubusercontent.com/180032/164207860-b6c9ee67-ce27-4504-926c-a825abc7a680.mp4

### Steps to reproduce

- Add DirectionalLight3D node with shadow enabled.
- Add a MeshInstance3D node with a BoxMesh acting as a ground (size: 640, 1, 640)
- Add a MeshInstance3D node with a BoxMesh acting as the long object (size: 1, 1, 640). Move it upwards so it can cast a shadow onto the floor.

### Minimal reproduction project

[test_shadow_long_object.zip](https://github.com/godotengine/godot/files/8520635/test_shadow_long_object.zip)

**Comment by mrjustaguy** [comment]
Increase subdivide (depth I think here) and it'll fix itself..

**Comment by Calinou** [comment]
I confirm increasing **Pancake Size** to 50 resolves the issue in the minimal reproduction project.

I'll leave this open as a documentation issue, so we don't forget to document this in the rewritten lights and shadows documentation.

**Comment by mrjustaguy** [comment]
Yeah it's probably not worth the effort trying to figure out exactly what's causing it and how to solve it (if it can be solved) when it seems it's just doing so because of too large triangles mixed with pancaking just going to hell.

In my observations (using Cubes) here's when this happens: (Multiplier > 1) * Face size > Pancake Size results in a possibility of the triangle failing to get shadows properly rendered in the shadow map.

**Comment by atirut-w** [comment]
I think this should be fixed as it would nerf open world projects since a *very* related issue breaks long distance shadows.

**Comment by Calinou** [comment]
> I think this should be fixed as it would nerf open world projects since a _very_ related issue breaks long distance shadows.

This issue is not fixable per se – it's a limitation of shadow pancaking to avoid shadow acne/peter-panning issues.

**Comment by atirut-w** [comment]
Are there other alternatives? I feel like not trying to find a solution to this conflicts with SDFGI's goal(GI solution for open world) since this would severely limit how big of a world you can make.

**Comment by mrjustaguy** [comment]
Ray Tracing.

But your statement that this severely limits how big of a world you can make is inaccurate.
You just need to make sure that your objects are sufficiently tessellated for each split, which in reality isn't often a problem

**Comment by atirut-w** [comment]
> You just need to make sure that your objects are sufficiently tessellated for each split, which in reality isn't often a problem

The models I used are sure as hell very tessellated: https://github.com/godotengine/godot/issues/63610#issuecomment-1198974201

**Comment by mrjustaguy** [comment]
Don't think that has anything to do with this issue, change to Orthogonal, and PCSS should not have those Splits visible..

**Comment by atirut-w** [comment]
> Don't think that has anything to do with this issue, change to Orthogonal, and PCSS should not have those Splits visible..

Huh, that worked, and the effect on quality isn't bad.

So this seems to be a separate problem? Has it already been reported?

**Comment by mrjustaguy** [comment]
I don't know if it has been reported, but it is indeed a separate problem.
